dev/test/index.ts

```diff
@@ -1273,3 +1296,389 @@ describe('getAll() method', () => {
     });
   });
 });
+
+describe('recursiveDelete() method:', () => {
+  // We store errors from batchWrite inside an error object since errors thrown
+  // in batchWrite do not affect the recursiveDelete promise.
+  let batchWriteError: Error | undefined;
+  let firestore: Firestore.Firestore;
+
+  beforeEach(() => {
+    setTimeoutHandler(setImmediate);
+  });
+
+  afterEach(() => {
+    verifyInstance(firestore);
+    setTimeoutHandler(setTimeout);
+  });
+
+  function instantiateInstance(
+    childrenDocs: Array<Array<string>>,
+    responses?: api.IBatchWriteResponse[]
+  ): Promise<Firestore.Firestore> {
+    let requestCount = 0;
+    const overrides: ApiOverride = {
+      runQuery: () => {
+        return stream(
+          ...childrenDocs[requestCount].map(docId => result(docId))
+        );
+      },
+      batchWrite: request => {
+        const expected = createRequest(
+          childrenDocs[requestCount].map(docId => deleteOp(docId))
+        );
+        try {
+          expect(request.writes).to.deep.equal(expected.writes);
+        } catch (e) {
+          batchWriteError = e;
```

Comment:
Do the tests not fail if we simply let the exception throw here? Ideally, the implementation should be able to handle if the `batchWrite` RPC throws an exception and propagate it as a failed BatchWrite.

Reply:
No, I think we can keep the logic as is. Individual RPC failures versus individual operations failures are not different from a user perspective. I was trying to see if we could simply bubble up the assertion failure as an individual RPC failure, which would surface these assertions in `onWriteResult` or via failed promises rather than via `batchWriteError`.